modules/rtdModule/index.js

```diff
@@ -159,130 +180,147 @@ export function init(config) {
   });
   events.on(CONSTANTS.EVENTS.AUCTION_END, (args) => {
-    subModules.forEach(sm => { sm.auctionEnd && sm.auctionEnd(args, sm.config) })
-  });
-  events.on(CONSTANTS.EVENTS.BEFORE_REQUEST_BIDS, (args) => {
-    subModules.forEach(sm => { sm.updateBidRequest && sm.updateBidRequest(args, sm.config) })
+    getAdUnitTargeting(args);
+    subModules.forEach(sm => { sm.onAuctionEndEvent && sm.onAuctionEndEvent(args, sm.config, _userConsent) })
   });
   events.on(CONSTANTS.EVENTS.BID_RESPONSE, (args) => {
-    subModules.forEach(sm => { sm.updateBidResponse && sm.updateBidResponse(args, sm.config) })
+    subModules.forEach(sm => { sm.onBidResponseEvent && sm.onBidResponseEvent(args, sm.config, _userConsent) })
   });
 }
 
 /**
- * get data from sub module
- * @param {AdUnit[]} adUnits received from auction
- * @param {function} callback callback function on data received
+ * loop through configured data providers If the data provider has registered getBidRequestData,
+ * call it, providing reqBidsConfigObj, consent data and module params
+ * this allows submodules to modify bidders
+ * @param {Object} reqBidsConfigObj required; This is the same param that's used in pbjs.requestBids.
+ * @param {function} fn required; The next function in the chain, used by hook.js
  */
-export function getProviderData(adUnits, callback) {
-  /**
-   * invoke callback if one of the conditions met:
-   * timeout reached
-   * all submodules answered
-   * all sub modules configured "waitForIt:true" answered (as long as there is at least one configured)
-   */
-
-  const waitForSubModulesLength = subModules.filter(sm => sm.config && sm.config.waitForIt).length;
-  let callbacksExpected = waitForSubModulesLength || subModules.length;
-  const shouldWaitForAllSubModules = waitForSubModulesLength === 0;
-  let dataReceived = {};
-  let processDone = false;
-  const dataWaitTimeout = setTimeout(done, _moduleConfig.auctionDelay || _moduleConfig.timeout || DEF_TIMEOUT);
-  subModules.forEach(sm => {
-    sm.getData(adUnits, onDataReceived.bind(sm));
+export function setBidRequestsData(fn, reqBidsConfigObj) {
+  initSubModules();
+
+  // delay bidding process only if auctionDelay > 0
+  if (!_moduleConfig.auctionDelay || _moduleConfig.auctionDelay < 1) {
+    return exitHook();
+  }
+
+  const relevantSubModules = subModules.filter(sm => typeof sm.getBidRequestData === 'function');
+  const prioritySubModules = relevantSubModules.filter(sm => !!(sm.config && sm.config.waitForIt));
+  let callbacksExpected = prioritySubModules.length;
+  let isDone = false;
+  let waitTimeout;
+
+  if (!relevantSubModules.length) {
+    return exitHook();
+  }
+
+  if (prioritySubModules.length) {
+    waitTimeout = setTimeout(exitHook, _moduleConfig.auctionDelay);
+  }
+
+  relevantSubModules.forEach(sm => {
+    sm.getBidRequestData(reqBidsConfigObj, onGetBidRequestDataCallback.bind(sm), sm.config, _userConsent)
```

Comment:
Shouldn't this support use case #2 in the [RTD module design doc](https://docs.google.com/document/d/1A9mPkH8PiuqmrtKmIao42_losnopwrCGLtLJPv3L4oM/edit?ts=5f623873#) ? 

> Update the AdUnits in the bidRequest object before going to the bidder (e.g. floors, first party data, adunit viewability)

Shouldn't the adUnits be passed here to allow the submodule to enrich them with data before the auction ?

Reply:
we are passing the global object - it will contain ad units if defined on request bids (and will be on the global ad units object if defined by `addAdunit`).
If you have a different suggestion I'll be happy to hear it

Reply:
@omerBrowsi I understand. 
How should we make our realTimeData available then to bid adapters when `buildRequests` is called ? Before, the rtd module would add them to a `realTimeData` object available in the `bidRequests` param. This is no longer the case, breaking the interface for bid adapters. Can we provide backwards compatibility for this ?

Reply:
There are only a couple of RTD sub-modules, we're not going to implement backwards compatibility.

Your RTD sub-mod needs to parse the reqBidsConfigObj and inject/remove whatever data you want.

Reply:
@omerBrowsi , @bretg and I discussed the absence of `realTimeData` on the `bidRequest` and we would like to add it back. I can add it back, but before I do so, Is there a reason it was removed ?

Reply:
I suspect it wasn't so much removed as the whole approach changed. It's entirely up to the RTD-submodule now -- RTD-core just passes the whole thing to each submodule where any change can be made. What we missed is that we should have set up a convention for each RTD-submodule to follow.